### PR TITLE
[Feat] Withdraw root constraint

### PR DESCRIPTION
**Pull request by noel2004**
This PR induce constraint for withdraw root, with following works planned:

* For bus-mapping:
  - [ ] ~~Add defination for `withdraw_root` and `withdraw_root_before` in `eth_block` (?)~~ Obtain `withdraw_root` and `withdraw_root_before` from sdb
  - [x] Induce l2's `message queue` precompile
  - [x] Add a read op from `message queue`'s `withdraw_root` slot in `end_block` step

Since we have no withdraw root avaliable yet, we can just use 0 in the circuit's witness. (Like reading from an unexisted contract)

* For circuit
  - [x] Add lookup for withdraw root reading in `end_block` gadget
  - [x] Export the assigned cell from evm circuit, just like state root being exported from state circuit
  - [x] Integrating into supercircuit: connect the exported cell to pi circuit
~~- [ ] *Temporarily* exclude this reading from mpt update until l2geth node support that (so the storage proof for withdraw slot is avaliable)~~

### CODE REVIEW

**Comment by noel2004** [comment]
Since the original code to connect `state_roots` with `pi_circuit` is not work (it try to immuate some fields in the circuit object), we have purposed another way to connect them.

**Comment by noel2004** [comment]
We should not include this PR until geth node has support (provide storage proof for message_queue precompile) this feature.

File: bus-mapping/src/precompile.rs [hunk 1]
```diff
@@ -3,6 +3,20 @@
...
+    use eth_types::U256;
+    use once_cell::sync::Lazy;
+    use std::str::FromStr;
+
+    /// address of L2MessageQueue predeploy
+    pub static MESSAGE_QUEUE: Lazy<Address> =
+        Lazy::new(|| Address::from_str("0x5300000000000000000000000000000000000000").unwrap());
+    /// the slot of withdraw root in L2MessageQueue
+    pub static WITHDRAW_TRIE_ROOT_SLOT: Lazy<U256> = Lazy::new(U256::zero);
+}
```
Comment (lispc) [inline bus-mapping/src/precompile.rs]: no i don't think MESSAGE_QUEUE can be called "precompile".   They can be named "l2_predeployed_contracts"

**Comment by lispc** [comment]
i would suggest add a check somewhere, to check withdraw_root_in_statedb_after_reexecution == withdraw_root_in_block_trace.

(checking this inside scroll-zkevm will also be ok)

similar to the "roots non consistent" log error

File: bus-mapping/src/precompile.rs (hunk 1 above)
Comment (noel2004) [reply bus-mapping/src/precompile.rs]: Now move the constants into `bus_mapping::l2_predeployed::message_queue`

**Comment by noel2004** [comment]
> i would suggest add a check somewhere, to check withdraw_root_in_statedb_after_reexecution == withdraw_root_in_block_trace.
> 
> (checking this inside scroll-zkevm will also be ok)
> 
> similar to the "roots non consistent" log error

Done.

We induced a method named `values` in `witness::mpt::MptUpdate` to expose the `old/new_value` memeber.

**Comment by noel2004** [comment]
Mock prove of supercircuit passed (baseon the alpha-testnet traces)

File: zkevm-circuits/src/evm_circuit/execution/end_block.rs [hunk 2]
```diff
@@ -131,6 +161,27 @@ impl<F: Field> ExecutionGadget<F> for EndBlockGadget<F> {
...
+
+        self.withdraw_root_assigned
+            .borrow_mut()
+            .replace(withdraw_root.cell());
+        // TODO: now we do not export withdraw_root_prev for we have only one
+        // phase2 cell which is enabled for copy constraint
+        // self.withdraw_root_prev_assigned
+        //     .borrow_mut()
+        //     .replace(withdraw_root_prev.cell());
+
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/end_block.rs:184]: is this still relevant? @noel2004

Comment (noel2004) [reply zkevm-circuits/src/evm_circuit/execution/end_block.rs:184]: Not any more. Current state should be ok

**Pull request merged**